### PR TITLE
switch-root: also check that mount IDs are the same, not just inodes

### PR DESCRIPTION
**Pull request by bluca**
If /run/nextroot/ has been set up, use it, even if the inodes are the same. It could be a verity device that is reused, but with different sub-mounts or other differences. Or the same / tmpfs with different /usr/ mounts. If it was explicitly set up we should use it irregardless of the inode check.

### CODE REVIEW

File: src/shared/switch-root.c [hunk 1]
```diff
@@ -66,11 +66,20 @@ int switch_root(const char *new_root,
...
         }
 
+        if (r == 0 && FLAGS_SET(flags, SWITCH_ROOT_DESTROY_OLD_ROOT)) {
+                istmp = fd_is_temporary_fs(old_root_fd);
+                if (istmp < 0)
+                        return log_error_errno(istmp, "Failed to stat root directory: %m");
+                if (istmp > 0)
+                        log_debug("Root directory is on tmpfs, will do cleanup later.");
+        } else
+                istmp = -1; /* don't know */
```
Comment (YHNdnzj) [inline src/shared/switch-root.c]: Please update the comment here.

File: src/basic/fd-util.c [hunk 2]
```diff
@@ -915,60 +912,78 @@ int path_is_root_at(int dir_fd, const char *path) {
...
+        STRUCT_NEW_STATX_DEFINE(st1);
+        STRUCT_NEW_STATX_DEFINE(st2);
+        int r;
+
+        assert(fd1 >= 0);
+        assert(fd2 >= 0);
+
+        r = statx_fallback(fd1, "", AT_EMPTY_PATH, STATX_TYPE|STATX_INO|STATX_MNT_ID, &st1.sx);
         if (r == -ENOTDIR)
                 return false;
```
Comment (poettering) [inline src/basic/fd-util.c]: why this check?

File: src/basic/fd-util.c [hunk 3]
```diff
@@ -915,60 +912,78 @@ int path_is_root_at(int dir_fd, const char *path) {
...
         if (r < 0)
                 return r;
 
-        r = statx_fallback(dir_fd, "..", 0, STATX_TYPE|STATX_INO|STATX_MNT_ID, &pst.sx);
+        r = statx_fallback(fd2, "", AT_EMPTY_PATH, STATX_TYPE|STATX_INO|STATX_MNT_ID, &st2.sx);
+        if (r == -ENOTDIR)
+                return false;
```
Comment (poettering) [inline src/basic/fd-util.c]: here too?

File: src/basic/fd-util.c (hunk 2 above)
Comment (bluca) [reply src/basic/fd-util.c]: dunno, it was there before so it just carried over, will remove

Comment (bluca) [reply src/basic/fd-util.c]: removed

**Review by yuwata — APPROVED** [review]

File: src/basic/fd-util.c (hunk 2 above)
Comment (poettering) [reply src/basic/fd-util.c]: it was there when it was referencing "." and "..", but now you reference "" with AT_EMPTY_PATH. The former only works on dirs, hence we needed to treat ENOTDIR. The latter works on any kind of inode, hence we don't.

Even after your rework you correctly retained the ENOTDIR check in path_is_root_at().

**Comment by poettering** [comment]
lgtm

**Pull request merged**